pkg/daemon/rpm-ostree.go

```diff
@@ -340,32 +330,107 @@ func (r *RpmOstreeClient) RebaseLayered(imgURL string) (changed bool, err error)
 	} else {
 		glog.Info("Current origin is not custom")
 	}
+}
 
-	glog.Infof("Executing rebase to %s", imgURL)
-	args := []string{"rebase", "--experimental", "ostree-unverified-registry:" + imgURL}
-
-	if err = runRpmOstree(args...); err != nil {
+// RebaseLayered potentially rebases system if not already rebased.
+func (r *RpmOstreeClient) RebaseLayered(imgURL string, pullSecret []byte) (err error) {
+	err = ioutil.WriteFile(ostreeAuthFile, pullSecret, 0400)
+	if err != nil {
 		return
 	}
+	defer os.Remove(ostreeAuthFile)
 
-	changed = true
-	return
+	glog.Infof("Executing rebase to %s", imgURL)
+	args := []string{"rebase", "--experimental", "ostree-unverified-registry:" + imgURL}
+
+	return runRpmOstree(args...)
 }
 
 // Live apply live-applies whatever we rebased to
 func (r *RpmOstreeClient) ApplyLive() (err error) {
 
 	glog.Infof("Applying live")
 
-	args := []string{"ex", "apply-live", "--allow-replacement"}
+	return runRpmOstree("ex", "apply-live", "--allow-replacement")
+}
 
-	if err = runRpmOstree(args...); err != nil {
-		return
+// allows unit testing
+var execCommand = exec.Command
+
+func runCmdSyncGetStdout(cmdName string, args ...string) (string, error) {
+	glog.Infof("Running: %s %s", cmdName, strings.Join(args, " "))
+	cmd := execCommand(cmdName, args...)
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		return "", errors.Wrapf(err, "error running %s %s: %s", cmdName, strings.Join(args, " "), string(stderr.Bytes()))
 	}
 
+	return string(stdout.Bytes()), nil
+}
+
+type FileTree struct {
+	children map[string]*FileTree
+}
+
+// ensures nodes exist for every element in a path
+func (node *FileTree) insert(path []string) {
+	child, ok := node.children[path[0]]
+	if !ok {
+		// if we're inserting the first child for this node we need to create the map
+		if node.children == nil {
+			node.children = make(map[string]*FileTree)
+		}
+		// there isn't a node for the first element in the path, so we need to create one
+		child = &FileTree{}
+		node.children[path[0]] = child
+	}
+	if len(path) > 1 {
+		child.insert(path[1:])
+	}
+}
+
+// walk the tree, generating a list of all paths to leaves
+func (node *FileTree) walk() (paths []string) {
+	for childPath, child := range node.children {
+		pathsToLeaves := child.walk()
+		if len(pathsToLeaves) == 0 {
+			// child is a leaf, so add it to paths
+			paths = append(paths, path.Join("/", childPath))
+		} else {
+			// else prepend childPath to all pathsToLeaves
+			for _, pathToLeaf := range pathsToLeaves {
+				paths = append(paths, path.Join("/", childPath, pathToLeaf))
+			}
+		}
+	}
 	return
 }
```

Comment:
The reason this happens is because ostree is content-addressed; each directory is represented by a checksum.  So any change to any file recursively changes every directory object up to the parent.

(Just like git)

So that's why we print out each directory as modified.

Reply:
Gotcha makes sense. Do you think generating a tree is a good approach or too complicated? I think @jkyros is okay with it but would prefer something easier to understand